api/v1alpha1/tls_types.go

```diff
@@ -0,0 +1,59 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+type TLSSettings struct {
+	// Version details the minimum/maximum TLS protocol verison that
+	// should be supported by this listener.
+	// +optional
+	Version *TLSVersions `json:"version,omitempty"`
+
+	// CipherSuites specifies the set of cipher suites supported when
+	// negotiating TLS 1.0 - 1.2. This setting has no effect for TLS 1.3.
+	// +optional
+	CipherSuites []string `json:"ciphers,omitempty"`
+
+	// ECDHCurves specifies the set of supported ECDH curves.
+	// +optional
+	ECDHCurves []string `json:"ecdhCurves,omitempty"`
+
+	// SignatureAlgorithms specifies which signature algorithms the listener should
+	// support.
+	// +optional
+	SignatureAlgorithms []string `json:"signatureAlgorithms,omitempty"`
+
+	// ALPNProtocols supplies the list of ALPN protocols that should be
+	// exposed by the listener. If left empty, ALPN will not be exposed.
+	// +optional
+	ALPNProtocols []string `json:"alpnProtocols,omitempty"`
```

Comment:
suggest a enumerate type instead of `string`, since [envoyproxy TLS](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/tls.proto.html#extensions-transport-sockets-tls-v3-commontlscontext) only support two values: `h2,http/1.1` or `http/1.1`.

IMO, we can add http protocol version enumerate type under `api/v1alpha1/shared_types.go`, like:
```go
HTTPProtocolVersion1_1 = "http/1.1"
HTTPProtocolVersion2   = "http/2"
```

and we can translate their combination into envoyproxy supported values:

- `["http/1.1", "http/2"]` will be seen as `h2,http/1.1`
- `["http/1.1"]` will be seen as `http/1.1`
- invalid otherwise

WDYT?

Reply:
I think maybe it's better to keep this as strings. I tried to create a configuration in Envoy where the ALPN was something else and Envoy had no issue with that - I was able to use `openssl s_client` to negotiate that different protocol string. 

If somebody uses a custom version of Envoy that is aware of a different protocol, it should be possible to configure it from EG.

Reply:
fair enough, since `alpn_protocols` is also a string type in envoyproxy

Reply:
+1, need to use a custom type to enforce cube builder validations
check out https://github.com/envoyproxy/gateway/blob/52946bef5132a302b6f5edd6d154062f35906e22/api/v1alpha1/ratelimit_types.go#L136

Reply:
But ALPN is more than just "h2" and "http/1.1". If I define this as an enum, it becomes impossible to use anything else. See @guydc's comment https://github.com/envoyproxy/gateway/pull/2287#discussion_r1425469498 below.

Reply:
do you have a specific use case where you'd like to set a specific ALPN apart from h2 and http/1.1 ?
its safer to use enums and validate synchronously imo

Reply:
My point is that validation the values to just "h2" and "http/1.1" is limiting. If this is moved to enums, I think it would make sense to have all of the possible values in the [IANA registry](https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#alpn-protocol-ids) available as accepted enums. 

EG is not there yet, but at some point it would make sense to have `ClientTrafficPolicy` documents attached also to TLS routes. When that happens, it would be possible to have other protocols supported. Examples would include SMTP, MQTT, or IMAP with EG terminating the TLS and routing to a service running in the cluster. If ALPN is limited to just the two possibilities - this won't be possible.

It honestly seems better to me to not set this as an enum. The default is set to `[ "h2", "http/1.1" ]`, meaning that most people wouldn't touch this. But when more control over this is needed, it makes more sense to me to make it possible to control this setting.

Reply:
Updated to be an enum, as agreed in the meeting.